src/coreclr/vm/pendingload.h

```diff
@@ -17,248 +17,404 @@
-    TypeHandle          m_typeHandle;
+        // Error result, propagated to all threads loading this class
+        HRESULT             m_hrResult;
 
- private:
-    // Type that we're loading
-    TypeKey             m_typeKey;
+        // Exception object to throw
+        Exception          *m_pException;
 
-    // Number of threads waiting for this type
-    LONG                m_dwWaitCount;
+        DWORD               m_dwHash;
 
-    // Error result, propagated to all threads loading this class
-    HRESULT             m_hrResult;
+        // m_Crst was acquired
+        bool                m_fLockAcquired;
 
-    // Exception object to throw
-    Exception          *m_pException;
+        bool                m_fIsPreallocated;
 
-    // m_Crst was acquired
-    BOOL                m_fLockAcquired;
-};
+        bool                m_fIsUnused = true;
+    };
 
-// Hash table used to hold pending type loads
-// @todo : use shash.h when it supports LoaderHeap/Alloc\MemTracker
-class PendingTypeLoadTable
-{
-protected:
-    struct TableEntry
+    struct Shard
     {
-        TableEntry* pNext;
-        DWORD                 dwHashValue;
-        PendingTypeLoadEntry* pData;
-    };
+        friend class PendingTypeLoadTable;
+        // This number chosen by experimentation with a fairly complex ASP.NET application that would naturally use about 40,000 Entry structures on startup.
+        // Entry allocations were shifted to about 11 during that startup phase.
+        static constexpr int PreallocatedEntryCount = 2;
+
+private:
+        Shard() : m_shardCrst(CrstUnresolvedClassLock)
+        {}
 
-    TableEntry     **m_pBuckets;    // Pointer to first entry for each bucket
-    DWORD           m_dwNumBuckets;
+        Entry *m_pLinkedListOfActiveEntries = NULL;
+        Crst                  m_shardCrst;
+        Entry  m_preAllocatedEntries[PreallocatedEntryCount];
 
 public:
+        Crst* GetCrst()
+        {
+            LIMITED_METHOD_CONTRACT;
+            return &m_shardCrst;
+        }
+
+        Entry* FindPendingTypeLoadEntry(DWORD hash, const TypeKey& typeKey)
+        {
+            WRAPPER_NO_CONTRACT;
+            for (PendingTypeLoadTable::Entry *current = m_pLinkedListOfActiveEntries; current != NULL; current = current->m_pNext)
+            {
+                if (current->m_dwHash != hash)
+                    continue;
+                TypeKey entryTypeKey = current->GetTypeKey();
+                if (TypeKey::Equals(&typeKey, &entryTypeKey))
+                {
+                    return current;
+                }
+            }
+
+            return NULL;
+        }
+
+        void RemovePendingTypeLoadEntry(Entry* pEntry)
+        {
+            LIMITED_METHOD_CONTRACT;
+
+            Entry **pCurrent = &m_pLinkedListOfActiveEntries;
+
+            while (*pCurrent != pEntry)
+            {
+                pCurrent = &((*pCurrent)->m_pNext);
+            }
+            *pCurrent = (*pCurrent)->m_pNext;
```

Comment:
I don't see the case being handled where `pEntry` isn't in the list.

Reply:
That isn't a valid case for calling this method. The `PendingTypeLoadEntry` *must* be in the list before calling this api. I'll add a comment.